worker/raft/rafttransport/worker.go

```diff
@@ -0,0 +1,336 @@
+
+package rafttransport
+
+import (
+	"context"
+	"log"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/raft"
+	"github.com/juju/errors"
+	"github.com/juju/loggo"
+	"github.com/juju/pubsub"
+	"gopkg.in/juju/names.v2"
+
+	"github.com/juju/juju/api"
+	"github.com/juju/juju/apiserver/apiserverhttp"
+	pubsubapiserver "github.com/juju/juju/pubsub/apiserver"
+	"github.com/juju/juju/worker/catacomb"
+	"github.com/juju/juju/worker/raft/raftutil"
+)
+
+var (
+	logger = loggo.GetLogger("juju.worker.raft.rafttransport")
+)
+
+// Config is the configuration required for running an apiserver-based
+// raft transport worker.
+type Config struct {
+	// APIInfo contains the information, excluding addresses,
+	// required to connect to an API server.
+	APIInfo *api.Info
+
+	// APIOpen is the function used to dial an API connection.
+	APIOpen api.OpenFunc
+
+	// Hub is the structured hub to which the worker will subscribe
+	// for API server address updates.
+	Hub *pubsub.StructuredHub
+
+	// Mux is the API server HTTP mux into which the handler will
+	// be installed.
+	Mux *apiserverhttp.Mux
+
+	// Path is the path of the raft HTTP endpoint.
+	Path string
+
+	// Tag is the tag of the agent running this worker.
+	Tag names.Tag
+
+	// Timeout, if non-zero, is the timeout to apply to transport
+	// operations. See raft.NetworkTransportConfig.Timeout for more
+	// details.
+	Timeout time.Duration
+}
+
+// Validate validates the raft worker configuration.
+func (config Config) Validate() error {
+	if config.APIInfo == nil {
+		return errors.NotValidf("nil APIInfo")
+	}
+	if config.APIOpen == nil {
+		return errors.NotValidf("nil APIOpen")
+	}
+	if config.Hub == nil {
+		return errors.NotValidf("nil Hub")
+	}
+	if config.Mux == nil {
+		return errors.NotValidf("nil Mux")
+	}
+	if config.Path == "" {
+		return errors.NotValidf("empty Path")
+	}
+	if config.Tag == nil {
+		return errors.NotValidf("nil Tag")
+	}
+	return nil
+}
+
+// NewWorker returns a new apiserver-based raft transport worker,
+// with the given configuration. The worker itself implements
+// raft.Transport.
+func NewWorker(config Config) (*Worker, error) {
+	if err := config.Validate(); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	w := &Worker{
+		config:         config,
+		connections:    make(chan net.Conn),
+		dialRequests:   make(chan dialRequest),
+		serversUpdated: make(chan struct{}),
+	}
+
+	const logPrefix = "[transport] "
+	logWriter := &raftutil.LoggoWriter{logger, loggo.DEBUG}
+	logLogger := log.New(logWriter, logPrefix, 0)
+	transport := raft.NewNetworkTransportWithConfig(&raft.NetworkTransportConfig{
+		Logger:                logLogger,
+		MaxPool:               7, // 7 is the maximum juju cluster size
+		ServerAddressProvider: serverAddressProvider{},
+		Stream: newStreamLayer(config.Tag, w.connections, &Dialer{
+			APIInfo: config.APIInfo,
+			DialRaw: w.dialRaw,
+			Path:    config.Path,
+		}),
+		Timeout: config.Timeout,
+	})
+	w.Transport = transport
```

Comment:
would it be clearer to construct transport first and then do create the worker struct and initialise all the fields in one go

Reply:
The stream layer depends on the worker's dialRaw method, so that at least needs to be constructed after the worker.